dsa/validate_test.go

```diff
@@ -80,99 +171,224 @@ func TestValidate(t *testing.T) {
 			},
 			giveBid: &entities.PbsOrtbBid{
 				Bid: &openrtb2.Bid{
-					Ext: json.RawMessage(`{"dsa": {}}`),
+					Ext: json.RawMessage(`{"dsa":{"behalf":"` + validBehalf + `","paid":"` + validPaid + `","adrender":2}}`),
+				},
+			},
+			wantError: false,
+		},
+		{
+			name: "required_and_bid_dsa_is_valid_no_adrender",
+			giveRequest: &openrtb_ext.RequestWrapper{
+				BidRequest: &openrtb2.BidRequest{
+					Regs: &openrtb2.Regs{
+						Ext: json.RawMessage(`{"dsa": {"dsarequired": 2, "pubrender": 0}}`),
+					},
+				},
+			},
+			giveBid: &entities.PbsOrtbBid{
+				Bid: &openrtb2.Bid{
+					Ext: json.RawMessage(`{"dsa":{"behalf":"` + validBehalf + `","paid":"` + validPaid + `"}}`),
 				},
 			},
-			wantValid: true,
+			wantError: false,
 		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			valid := Validate(tt.giveRequest, tt.giveBid)
-			assert.Equal(t, tt.wantValid, valid)
+			err := Validate(tt.giveRequest, tt.giveBid)
+			if tt.wantError {
+				assert.Error(t, err)
+			} else {
+				assert.Nil(t, err)
+			}
 		})
 	}
 }
 
 func TestDSARequired(t *testing.T) {
 	tests := []struct {
 		name         string
-		giveRequest  *openrtb_ext.RequestWrapper
+		giveReqDSA   *openrtb_ext.ExtRegsDSA
 		wantRequired bool
 	}{
 		{
-			name: "not_required_and_reg.ext.dsa_is_nil",
-			giveRequest: &openrtb_ext.RequestWrapper{
-				BidRequest: &openrtb2.BidRequest{
-					Regs: &openrtb2.Regs{
-						Ext: json.RawMessage(`{}`),
-					},
-				},
+			name:         "nil",
+			giveReqDSA:   nil,
+			wantRequired: false,
+		},
+		{
+			name: "nil_required",
+			giveReqDSA: &openrtb_ext.ExtRegsDSA{
+				Required: nil,
+			},
+			wantRequired: false,
+		},
+		{
+			name: "not_required_0",
```

Comment:
Nitpick: Do you think we should replace the numbers in these test names with the name condition being tested?